flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperator.java

```diff
@@ -477,7 +478,7 @@ public void initializeState(StateInitializationContext context) throws Exception
 				for (int i = 0; i < noOfTimerServices; i++) {
 					String serviceName = div.readUTF();
 
-					HeapInternalTimerService<?, ?> timerService = this.timerServices.get(serviceName);
+					InternalTimerService<?, ?> timerService = this.timerServices.get(serviceName);
```

Comment:
Maybe you could add a parameter to the factory method that allow us to derive the generic types of key and value (e.g. the type serializers).